Release v0.6.0

jasonbahl opened the pull request:
## WIP

This PR is a Work in Progress, but aims to solve quite a few issues with this plugin. 

## Breaking Changes

This PR includes many breaking changes, both to internal code APIs for users extending code via WordPress hooks/filters, etc _and_ breaking changes to the shape of the Schema, so client applications will need to be adjusted to compensate for the changes.

## Issues Addressed

This addresses the issues in the v0.6.0 milestone: https://github.com/wp-graphql/wp-graphql-acf/milestone/2

I'll be going through each issue and commenting/documenting how this PR resolves the issue. 

- closes [#173](https://github.com/wp-graphql/wp-graphql-acf/issues/173#issuecomment-840677785)
- closes [#167](https://github.com/wp-graphql/wp-graphql-acf/issues/167#issuecomment-840803289)
- closes [#243](https://github.com/wp-graphql/wp-graphql-acf/issues/243#issuecomment-840685024)
- closes [#247](https://github.com/wp-graphql/wp-graphql-acf/issues/247#issuecomment-840754119)
- closes [#237](https://github.com/wp-graphql/wp-graphql-acf/issues/237#issuecomment-840760327)
- closes [#164](https://github.com/wp-graphql/wp-graphql-acf/issues/164#issuecomment-840764451)
- closes [#16](https://github.com/wp-graphql/wp-graphql-acf/issues/16#issuecomment-840791179)
- closes [#85](https://github.com/wp-graphql/wp-graphql-acf/issues/85#issuecomment-842444795)

jasonbahl committed:
- Refactor Fields, resolve field group prefixing issues, re-use of cloned field groups
- deprecate class-config.php (will be removed for release)

jasonbahl committed:
no message

jasonbahl committed:
Merge branch 'develop' into feature/fix-field-group-prefixing

# Conflicts:
#	src/deprecated-class-config.php
#	vendor/autoload.php
#	vendor/composer/autoload_real.php
#	vendor/composer/autoload_static.php

jasonbahl committed:
- Adjust the JS on the GraphQL Settings to only execute the AJAX request if one is not already pending.

jasonbahl committed:
Merge branch 'release/v0.5.0' into feature/fix-field-group-prefixing

# Conflicts:
#	src/js/main.js

jasonbahl committed:
- Add PHPStan checks
- Start adding PHPStan suggested fixes
- BREAKING: Changed some class names

jasonbahl committed:
- remove old (renamed) class names

jasonbahl committed:
Merge commit '49bf9558508d6167d0b3795df8798641b95b06e2' into feature/fix-field-group-prefixing

# Conflicts:
#	vendor/composer/InstalledVersions.php
#	vendor/composer/installed.php

jasonbahl committed:
Merge branch 'bug/#251-field-groups-not-in-graphql-are-preventing-other-groups-from-showing' into feature/fix-field-group-prefixing

jasonbahl committed:
- Add test for: https://github.com/wp-graphql/wp-graphql-acf/issues/173

Merge branch 'develop' into feature/fix-field-group-prefixing

# Conflicts:
#	vendor/composer/InstalledVersions.php
#	vendor/composer/installed.php

ardiewen commented:
Understanding that this is still very much WIP, the major question that still remains open for me is how do we address reusability of ACF Fields on both the content editor side, as well as the graphql schema.

The following simplified use case will hopefully illustrate the need and the current gap in functionality.

<details><summary>Context / Use Case</summary>

As a layout builder, we've found great utility in building standalone field groups that address certain functions, then cloning them as needed. Here's a simplified / contrived example:

I've created a custom field group called `Background` with a single button group field called "Kind" which has options ["color", "image", "video"].

<img width="936" alt="Screen Shot 2021-05-13 at 3 48 25 PM" src="https://user-images.githubusercontent.com/39966887/118178970-ab314800-b402-11eb-992d-7760757e167b.png">

We can then create our first section, called `Banner` and clone the entire field group using the `All fields from...` option.

<img width="716" alt="Screen Shot 2021-05-13 at 3 52 49 PM" src="https://user-images.githubusercontent.com/39966887/118179432-488c7c00-b403-11eb-9a48-7bbd27b81112.png">

When we create addition sections such as `Accordion`, `Carousel`, `Columns`, in the future, any upstream changes to the `Background` field group will automatically be available on these 
</details>

Currently, this is what we're observing on this PR branch:

1. No prefixes are applied automatically to field groups when they're assigned using location rules.

   This is great! From our contrived example above, it means we can create a fragment like this:

   ```
   fragment BackgroundFields on Background {
     kind
   }
   ```

2. When field groups are cloned however, they automatically inherit the parent type 

   This becomes a challenge, as you'll get fragments like this:
   
   <img width="310" alt="Screen Shot 2021-05-13 at 4 08 06 PM" src="https://user-images.githubusercontent.com/39966887/118181163-6c50c180-b405-11eb-8ff2-9ff2ce6e833f.png">

   Which means we **_cannot_** reuse our fragment above:

   ```
   fragment BannerFields on Banner {
     background {
       ...BackgroundFields
     }
   }
   ```

   Ideally, we'd have a scenario in which we can clone the `Background` field group into multiple layout field groups, while reusing the same fragment above, like so:

   ```
   fragment BannerFields on Banner {
     background {
       ...BackgroundFields
     }
   }
   fragment AccordionFields on Accordion {
     background {
       ...BackgroundFields
     }
   }
   fragment CarouselFields on Carousel {
     background {
       ...BackgroundFields
     }
   }
   ...
   ```

I'm not sure how much of this is by design vs due to WIP status, but I thought I'd report some initial findings.

There's also a lot of this that hinges upon the design / intent of the clone field, I suppose.

Anyway, if there's anything else you'd like me to test, happy to make some more examples.

jasonbahl committed:
- Update test for gallery to reflect proper order
- add test for taxonomy field in repeaters
- add test for repeater field with no values saved

jasonbahl committed:
- Fixes bug where querying a flex field with no data throws errors

jasonbahl committed:
- refactor registry and fields.
- all tests passing. . .still need to work on Clone fields, Flex Fields, and re-usable interfaces/fragments. . . and need to clean up code quite a bit

jasonbahl committed:
- Update WPGraphQL Test Case to v2.0
- Add initial work to open Field Group as a Fragment in GraphiQL (in progress, but shows where in the Schema Field Groups will show in the Field Group posts table)
- Add helper function `get_parent_type_fields_interface()` to AcfField.php
- Update fields/FlexibleContent.php to register Layout Interfaces and return the Flex Field type as a list of the Layout Interface
- Update Connection Fields to use the Field Group Interface type as the to-type instead of the Field Group Object Type
- Add a complex field group export for tests
- Add initial tests for Settings (testing the fragment generation)
- Add initial tests for ComplexFlexAndCloneFields

jasonbahl committed:
- scaffold out files for docs

jasonbahl committed:
Move docs for each field type out of the main README.md and into their own files.

No changes to the content of the docs, just organization of the docs.

jasonbahl committed:
Merge branch 'feature/refactor-docs' into feature/fix-field-group-prefixing

# Conflicts:
#	docs/fields/accordion.md
#	docs/fields/button-group.md
#	docs/fields/text.md

jasonbahl committed:
- Updating composer dependencies
- run `composer install --no-dev`

gavinhenderson commented:
Hey @jasonbahl @kidunot89 Is there any plan to release this new version?

If there is anything that I can contribute to help you be able to launch this let me know. I am very appreciative of all your work!

I am in need of some of the fixes in this new version to allow me to access relationship fields that is a list.

jasonbahl commented:
@gavinhenderson I'm hoping to get back to it soon, but I can't promise an ETA as there's a lot of other in-flight work I'm doing.

jasonbahl commented:
But feel free to use this branch in your environments if it works for you!
